tensorflow/core/graph/gpu_fusion_pass.cc

```diff
@@ -0,0 +1,710 @@
+    VLOG(lvl) << "\t" << n;
+  }
+  VLOG(lvl) << "===========";
+}
+
+bool isGpuDevice(StringPiece fullname) {
+  DeviceNameUtils::ParsedName p;
+  return (DeviceNameUtils::ParseFullName(fullname, &p) &&
+          (p.type == kGPUDeviceStr));
+}
+
+bool areAssignedToSameGpu(std::list<const Node*> nodes) {
+  int device_id = -1;
+  bool on_same_gpu = true;
+
+  DeviceNameUtils::ParsedName p;
+  bool first_node = true;
+  for (auto n : nodes) {
+    if (!n->has_assigned_device_name()) {
+      on_same_gpu = false;
+      break;
+    }
+
+    DeviceNameUtils::ParseFullName(n->assigned_device_name(), &p);
+    if (p.type != kGPUDeviceStr) {
+      on_same_gpu = false;
+      break;
+    }
+
+    if (first_node) {
+      device_id = p.id;
+    } else if (device_id != p.id) {
+      on_same_gpu = false;
+      break;
+    }
+  }
+
+  return on_same_gpu;
+}
+
+// is this node an instance of a convolution op for which we support fusion for?
+inline bool isOpConvolution(const Node* n) {
+  return (n->type_string() == "Conv2D");
+}
+
+// is this node an instance of a bias op for which we support fusion for?
+inline bool isOpBias(const Node* n) { return (n->type_string() == "BiasAdd"); }
+
+// is this node an instance of a activation op for which we support fusion for?
+inline bool iaOpActivation(const Node* n) {
+  return ((n->type_string() == "Sigmoid") || (n->type_string() == "Relu") ||
+          (n->type_string() == "Relu6") || (n->type_string() == "Tanh"));
+}
+
+// is this node an instance of a batchnorm op for which we support fusion for?
+inline bool isOpBatchNorm(const Node* n) {
+  return ((n->type_string() == "FusedBatchNorm") ||
+          (n->type_string() == "FusedBatchNormV2"));
+}
+// Util routines - End
+
+//----------------------------------------------------------------------
+
+class ROCmFusionOpBase;
+
+class ROCmFusionPass : public GraphOptimizationPass {
+ public:
+  // optimization pass entry point,
+  // application code will call this routine to run the pass
+  Status Run(const GraphOptimizationPassOptions& options) override;
+
+  // helper function that does all the work for this pass
+  bool RunPass(Graph* g);
+
+ private:
+  void InitializeFusions(
+      std::vector<std::unique_ptr<ROCmFusionOpBase> >& fusions, Graph* g,
+      std::set<const Node*>& fused_nodes);
+
+};
+
+// Register the ROCmFusionPass with the registry.
+// The choice of phase number (1) is arbitrary.
+
+REGISTER_OPTIMIZATION(kROCmFusionPassGrouping,  // grouping
+                      1,                        // phase number
+                      ROCmFusionPass);
+
+//----------------------------------------------------------------------
+
+// absract base class for an individual fusion operation
+class ROCmFusionOpBase {
+ public:
+  ROCmFusionOpBase(Graph* g, std::set<const Node*>& fn, string n)
+      : graph_(g), fused_nodes_(fn), fusion_op_name_(n) {}
+
+  virtual ~ROCmFusionOpBase() {}
+
+  virtual bool DoFusion(const Node* n) = 0;
```

Comment:
@deven-amd also please use `Status` to replace `bool` so this pass could be upstreamed in the near future.

Reply:
`DoFusion` is only used within this file. 

The only routine (in this file) that can be called by code outside of this file is `ROCmFusionPass::Run` and that does have a `Status` return value.

I would prefer to leave the return value of `DoFusion` as `bool` given that it signifies whehther or not the routine applied fusion to the given node. I will add a comment to the same effect.

Given this, is it ok to leave the return value of `DoFusion` as `bool`?

Reply:
thanks for the clarification. if `DoFusion` is an internal member function used solely within this module i think it's okay to leave it to return `bool`.